torch/_guards.py

```diff
@@ -645,6 +645,24 @@ class ChainedSource(Source):
     base: Source
 
 
+# Represents a source tensor coming from a field on a wrapper tensor subclass.
+# For example, in torch.testing._internal.double_tensor.DoubleTensor,
+# fields DoubleTensor.a and DoubleTensor.b will get a WrapperSubclassFieldSource
+@dataclasses.dataclass(frozen=True)
+class WrapperSubclassFieldSource(Source):
```

Comment:
Maybe we should still have a wrapper specific source, but it's not this one.

I can think of two coherent designs.

1. We fiat that all tensor subclasses MUST store their tensors as plain attributes that can be accessed with getattr. `__tensor_flatten__` returns a list of attributes to poke at to get at tensors (it does NOT return the actual tensors, because remember, we can always use getattr to get at the tensors). You don't need a WrapperSubclassFieldSource, you can just use AttrSource
2. The tensor subclasses may store tensors anywhere, and `__tensor_flatten__` is the only way you can reliably extract them. Then, if we want to test on an inner tensor, we must run `__tensor_flatten__` and then index out the particular tensor you want. You need a custom source for this. You need a fairly complicated reconstruct implementation.

(1) is more efficient but you need to change the `__tensor_flatten__` calling convention.

Reply:
This makes sense. I was effectively trying to do (1) here, but didn't realize that I could just use an existing `AttrSource`.

It sounds like the risk of (1) is that... we're relying on them implementing `__tensor_flatten__` correctly or they risk incorrect guards, and in particular you pointed out that this requires every tensor to have an actual attr (and not be e.g. stashed in a list. This seems.. probably reasonable, with documentation?)

Reply:
I'll read through the `Source` code some more to understand how to re-use `AttrSource`, and update `__tensor_flatten__` to only return list-of-strings.

Reply:
Any preference on what we should pass to `__tensor_unflatten__`? The default I would probably choose is that:
- our code calls __tensor_flatten__() to get a list of attributes
- our code grabs each attribute off the tensor and transforms it in some way
- our code constructs a dict of (attribute_str, updated_tensor) and passes that dictto `__tensor_unflatten__`. Maybe we could just pass the tensors as a list instead of a dict keyed on attribute, but that doesn't clearly tell the subclass which tensor is which.

Reply:
I think a dict of attribute to updated tensor is still OK for unflatten

Reply:
this is now done